From e174eb3a9adf9a838068b9f184f4e648f553bd4b Mon Sep 17 00:00:00 2001
From: Ernesto Alfonso <erjoalgo@gmail.com>
Date: Wed, 30 Aug 2023 17:26:41 -0400
Subject: [PATCH] Add support for the special ${RANDOM} environment variable.

This can be used to generate strings unique to each ssh connection.
For example, it can be used to generate unique RemoteForward unix
socket paths to avoid conflicts between multiple connections to a given
remote host.
---
 misc.c       | 14 ++++++++++++--
 ssh_config.5 |  2 ++
 2 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/misc.c b/misc.c
index c75a795c..2eb51dae 100644
--- a/misc.c
+++ b/misc.c
@@ -1110,6 +1110,7 @@ vdollar_percent_expand(int *parseerror, int dollar, int percent,
 	int r, missingvar = 0;
 	char *ret = NULL, *var, *varend, *val;
 	size_t len;
+	char random[7];
 
 	if ((buf = sshbuf_new()) == NULL)
 		fatal("%s: sshbuf_new failed", __func__);
@@ -1152,8 +1153,17 @@ vdollar_percent_expand(int *parseerror, int dollar, int percent,
 			}
 			var = xmalloc(len + 1);
 			(void)strlcpy(var, string, len + 1);
-			if ((val = getenv(var)) == NULL) {
-				error("%s: env var ${%s} has no value",
+			val = NULL;
+			if (strcmp(var, "RANDOM") == 0) {
+				/*
+				 * Generate a random 5-digit number to support the
+				 * special ${RANDOM} environment variable.
+				 */
+				snprintf(random, sizeof(random), "%d", arc4random_uniform(1e6));
+				val = random;
+			}
+			if (val == NULL && (val = getenv(var)) == NULL) {
+                            error("%s: env var ${%s} has no value",
 				    __func__, var);
 				missingvar = 1;
 			} else {
diff --git a/ssh_config.5 b/ssh_config.5
index 7f64c2cf..23df6679 100644
--- a/ssh_config.5
+++ b/ssh_config.5
@@ -2215,6 +2215,8 @@ for example
 would refer to the user's .ssh directory.
 If a specified environment variable does not exist then an error will be
 returned and the setting for that keyword will be ignored.
+The special environment variable ${RANDOM} is supported and expands to a
+random non-negative number of at most 5 digits.
 .Pp
 The keywords
 .Cm CertificateFile ,
-- 
2.39.2

